crates_io_cdn_logs/src/paths.rs

```diff
@@ -0,0 +1,113 @@
+use semver::Version;
+use tracing::instrument;
+
+/// Parse crate name and version from a download URL or URL path.
+#[instrument(level = "debug")]
+pub fn parse_path(mut path: &str) -> Option<(String, Version)> {
+    // This would ideally use a regular expression to simplify the code, but
+    // regexes are slow, and we want to keep this code as fast as possible.
+
+    // Remove any query parameters.
+    if let Some(pos) = path.find('?') {
+        path = &path[..pos];
+    }
+
+    // Find the start of the path. We assume that we don't have any nested
+    // `crates` folders on the server (e.g. `/foo/crates/...`).
+    let pos = path.find("/crates/")?;
+    let path = &path[pos + 8..];
+
+    let (folder, filename) = path.split_once('/')?;
+    let filename = filename.strip_suffix(".crate")?;
+
+    let version = filename.strip_prefix(folder)?;
+    let version = version.strip_prefix('-')?;
+    let version = Version::parse(version).ok()?;
+
+    Some((folder.to_owned(), version))
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I wouldn't mind seeing a couple more tests for these specific cases:

1. A crate name with one or more dashes, eg `/crates/foo-bar/foo-bar-1.0.0.crate`.
2. A version number with one or more dashes, eg `/crates/foo/foo-1.0.0-rc.1.crate`.

(I know these are handled correctly now, but just in case someone decides to "improve" this later.)